components/sup/src/manager/service/spec.rs

```diff
@@ -62,66 +106,50 @@ impl ServiceSpec {
         if self.ident == PackageIdent::default() {
             return Err(sup_error!(Error::MissingRequiredIdent));
         }
-
-        toml::to_string(self).map_err(|e| sup_error!(Error::ServiceSpecRender(e.to_string())))
+        toml::to_string(self).map_err(|err| sup_error!(Error::ServiceSpecRender(err)))
     }
 
     pub fn from_file<P: AsRef<Path>>(path: P) -> Result<Self> {
-        let file = File::open(&path).map_err(|e| {
-                         sup_error!(Error::ServiceSpecFileRead(path.as_ref().display().to_string(),
-                                                               e.to_string()))
-                     })?;
+        let file =
+            File::open(&path).map_err(|err| {
+                             sup_error!(Error::ServiceSpecFileIO(path.as_ref().to_path_buf(), err))
+                         })?;
         let mut file = BufReader::new(file);
         let mut buf = String::new();
         file.read_to_string(&mut buf)
-            .map_err(|e| {
-                         sup_error!(Error::ServiceSpecFileRead(path.as_ref().display().to_string(),
-                                                               e.to_string()))
-                     })?;
-
+            .map_err(|err| sup_error!(Error::ServiceSpecFileIO(path.as_ref().to_path_buf(), err)))?;
         Self::from_str(&buf)
     }
 
     pub fn to_file<P: AsRef<Path>>(&self, path: P) -> Result<()> {
         debug!("Writing service spec to '{}': {:?}",
                path.as_ref().display(),
                &self);
-        let dst_path = path.as_ref()
-            .parent()
-            .ok_or(sup_error!(Error::ServiceSpecFileWrite(path.as_ref().display().to_string(),
-                                                          "cannot determine parent directory"
-                                                              .to_string())))?;
+        let dst_path =
+            path.as_ref().parent().expect("Cannot determine parent directory for service spec");
```

Comment:
Input validation is done at a higher level than this. We do actually mean to panic here because we're communicating back to the developer that they fucked something up if we ever hit this